File: frontend/src/api/notify.ts
import { request } from '@/utils/request'
import type { ApiResponse } from '@/types/common'

export interface NotifyItem {
  id: number
  title: string
  content: string
  type: string // system / feature / maintenance
  level: string // info / warning / error / success
  status: number
  start_at?: string
  end_at?: string
  created_at: string
  updated_at: string
}

export interface NotifyListData {
  items: NotifyItem[]
}

export const notifyApi = {
  // 获取当前活跃通知（公开）
  active: (type?: string) => {
    return request<ApiResponse<NotifyListData>>({
      url: '/notifies/active',
      method: 'GET',
      params: type ? { type } : undefined,
    })
  },

  // 列表（管理）
  list: (params: { page?: number; page_size?: number; type?: string; level?: string; status?: number }) => {
    return request<ApiResponse<{
      items: NotifyItem[]
      total: number
      page: number
      page_size: number
    }>>({
      url: '/admin/notifies',
      method: 'GET',
      params,
    })
  },

  get: (id: number) => {
    return request<ApiResponse<NotifyItem>>({
      url: `/admin/notifies/${id}`,
      method: 'GET',
    })
  },

  create: (data: Omit<NotifyItem, 'id' | 'created_at' | 'updated_at' | 'status'> & { status?: number }) => {
    return request<ApiResponse<NotifyItem>>({
      url: '/admin/notifies',
      method: 'POST',
      data,
    })
  },

  update: (id: number, data: Partial<NotifyItem>) => {
    return request<ApiResponse<unknown>>({
      url: `/admin/notifies/${id}`,
      method: 'PUT',
      data,
    })
  },

  delete: (id: number) => {
    return request<ApiResponse<unknown>>({
      url: `/admin/notifies/${id}`,
      method: 'DELETE',
    })
  },
}

File: backend/internal/repo/db/model/notify.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// Notify 通知公告
type Notify struct {
	gorm.Model
	Title        string     `gorm:"size:255;not null" json:"title"`
	Content      string     `gorm:"type:text;not null" json:"content"`
	Type         string     `gorm:"size:20;default:'system';index" json:"type"` // system/feature/maintenance/share_retrieved
	Level        string     `gorm:"size:20;default:'info';index" json:"level"`  // info/warning/error/success
	Status       int        `gorm:"default:1;index" json:"status"`              // 0=草稿 1=发布 2=下线
	StartAt      *time.Time `json:"start_at,omitempty"`                         // 生效时间
	EndAt        *time.Time `json:"end_at,omitempty"`                           // 失效时间
	AuthorID     uint       `gorm:"index" json:"author_id"`                     // 发布管理员
	TargetUserID *uint      `gorm:"index" json:"target_user_id,omitempty"`      // 0=广播, >0=定向给某用户
	ReadAt       *time.Time `gorm:"index" json:"read_at,omitempty"`             // 用户首次已读时间
}

// TableName 自定义表名
func (Notify) TableName() string {
	return "notifies"
}

// IsActive 当前是否活跃（在时间窗口内 + 状态发布）
func (n *Notify) IsActive() bool {
	if n.Status != 1 {
		return false
	}
	now := time.Now()
	if n.StartAt != nil && now.Before(*n.StartAt) {
		return false
	}
	if n.EndAt != nil && now.After(*n.EndAt) {
		return false
	}
	return true
}
